src/sentry/api/endpoints/organization_dashboard_details.py

```diff
@@ -1,11 +1,89 @@
 from __future__ import absolute_import
 
+from django.db import IntegrityError, transaction
+from rest_framework import serializers
+from rest_framework.response import Response
+
 from sentry.api.base import DocSection
 from sentry.api.bases.dashboard import (
     OrganizationDashboardEndpoint
 )
 from sentry.api.serializers import serialize
-from sentry.models import ObjectStatus
+from sentry.api.serializers.rest_framework import get_next_dashboard_order, ListField, ValidationError
+from sentry.models import ObjectStatus, Widget
+
+
+def remove_widgets(dashboard_widgets, widget_data):
+    """
+    Removes current widgets belonging to dashboard not in widget_data.
+    Returns remaining widgets.
+    """
+    widget_ids = [wd['id'] for wd in widget_data]
+    dashboard_widgets.exclude(
+        id__in=widget_ids
+    ).delete()
+    return dashboard_widgets.filter(
+        id__in=widget_ids
+    )
+
+
+def reorder_widgets(dashboard_id, widget_data):
+    """
+    Reorders Widgets given the relative order desired,
+    reorders widgets in the next possible set of numbers
+    i.e if order of widgets is 1, 2, 3
+        the reordered widgets will have order 4, 5, 6
+    """
+    dashboard_widgets = Widget.objects.filter(
+        dashboard_id=dashboard_id,
+    )
+    dashboard_widgets = list(remove_widgets(dashboard_widgets, widget_data))
+
+    # dashboard_widgets and widget_data should now have the same widgets
+    widget_data.sort(key=lambda x: x['order'])
+
+    next_order = get_next_dashboard_order(dashboard_id)
+    for index, data in enumerate(widget_data):
+        for widget in dashboard_widgets:
+            if widget.id == data['id']:
+                widget.order = next_order + index
+                widget.save()
+                break
+
+
+class WidgetSerializer(serializers.Serializer):
+    order = serializers.IntegerField(min_value=0, required=True)
+    id = serializers.IntegerField(min_value=0, required=True)
+
+
+class DashboardWithWidgetsSerializer(serializers.Serializer):
+    title = serializers.CharField(required=False)
+    widgets = ListField(
+        child=WidgetSerializer(),
+        required=False,
+        allow_null=True,
+    )
+
+    def validate_widgets(self, attrs, source):
+        try:
+            widgets = attrs[source]
+        except KeyError:
+            return attrs
+
+        if len(widgets) != len(set([w['order'] for w in widgets])):
+            raise ValidationError('Widgets must not have duplicate order values.')
+
+        widgets_count = len(Widget.objects.filter(
```

Comment:
@markstory is it ok that I have this in here?

Reply:
I think so, but this will hydrate all the records and then discard them after counting them. You could use `objects.filter(...).count()` to get just the integer value back from the database.